pkg/cloudprovider/instance.go

```diff
@@ -52,17 +51,28 @@ var (
 	instanceTypeFlexibilityThreshold = 5 // falling back to on-demand without flexibility risks insufficient capacity errors
 )
 
-type InstanceProvider struct {
+// InstanceProvider enables creating, getting and deleting EC2 instances based on Karpenter specs
+type InstanceProvider interface {
+	Create(ctx context.Context, nodeTemplate *v1alpha1.AWSNodeTemplate, machine *v1alpha5.Machine, instanceTypes []*cloudprovider.InstanceType) (*ec2.Instance, error)
+	Get(ctx context.Context, id string) (*ec2.Instance, error)
+	Delete(ctx context.Context, machine *v1alpha5.Machine) error
+}
+
+type instanceProviderImpl struct {
```

Comment:
I'm not a fan of the `Impl` suffix, but I'm struggling to come up with something better since I generally prefer the `er` suffix on the interface and then without it on the implementation, but that doesn't really work here...